transports/webrtc/src/transport.rs

```diff
@@ -0,0 +1,784 @@
+        // safe to unwrap here because we require a certificate during construction.
+        let fingerprints = self
+            .inner
+            .certificates
+            .first()
+            .expect("at least one certificate")
+            .get_fingerprints()
+            .expect("get_fingerprints to succeed");
+        // TODO:
+        //   modify webrtc-rs to return value in upper-hex rather than lower-hex
+        //   Fingerprint::from(fingerprints.first().unwrap())
+        debug_assert_eq!("sha-256", fingerprints.first().unwrap().algorithm);
+        Fingerprint::new_sha256(fingerprints.first().unwrap().value.to_uppercase())
+    }
+
+    /// Consumes the `WebRTCConfiguration`, returning its inner configuration.
+    pub fn into_inner(self) -> RTCConfiguration {
+        self.inner
+    }
+}
+
+/// Turns an IP address and port into the corresponding WebRTC multiaddr.
+pub(crate) fn socketaddr_to_multiaddr(socket_addr: &SocketAddr) -> Multiaddr {
+    Multiaddr::empty()
+        .with(socket_addr.ip().into())
+        .with(Protocol::Udp(socket_addr.port()))
+        .with(Protocol::WebRTC)
+}
+
+/// Extracts a SHA-256 fingerprint from the given address. Returns `None` if the address does not
+/// contain one.
+fn fingerprint_from_addr(addr: &Multiaddr) -> Option<Fingerprint> {
+    let iter = addr.iter();
+    for proto in iter {
+        match proto {
+            // Only support SHA-256 (0x12) for now.
+            Protocol::Certhash(f) if f.code() == 0x12 => return Some(Fingerprint::from(f)),
+            _ => continue,
+        }
+    }
+    None
+}
+
+/// Tries to turn a WebRTC multiaddress into a [`SocketAddr`]. Returns None if the format of the
+/// multiaddr is wrong.
+fn multiaddr_to_socketaddr(addr: &Multiaddr) -> Option<SocketAddr> {
+    let mut iter = addr.iter();
+    let proto1 = iter.next()?;
+    let proto2 = iter.next()?;
+    let proto3 = iter.next()?;
+
+    // Return `None` if protocols other than `p2p` or `certhash` are present.
+    for proto in iter {
+        match proto {
+            Protocol::P2p(_) => {}
+            Protocol::Certhash(_) => {}
+            _ => return None,
+        }
+    }
+
+    match (proto1, proto2, proto3) {
+        (Protocol::Ip4(ip), Protocol::Udp(port), Protocol::WebRTC) => {
+            Some(SocketAddr::new(ip.into(), port))
+        }
+        (Protocol::Ip6(ip), Protocol::Udp(port), Protocol::WebRTC) => {
+            Some(SocketAddr::new(ip.into(), port))
+        }
+        _ => None,
+    }
+}
+
+async fn perform_noise_handshake_outbound<T>(
+    id_keys: identity::Keypair,
+    poll_data_channel: T,
+    our_fingerprint: Fingerprint,
+    remote_fingerprint: Fingerprint,
+) -> Result<PeerId, Error>
+where
+    T: AsyncRead + AsyncWrite + Unpin + Send + 'static,
+{
+    let dh_keys = Keypair::<X25519Spec>::new()
+        .into_authentic(&id_keys)
+        .unwrap();
+    let noise = NoiseConfig::xx(dh_keys);
```

Comment:
Once https://github.com/libp2p/rust-libp2p/pull/2903 is merged, this is where we need to use the new `with_prologue` API: https://github.com/libp2p/rust-libp2p/pull/2903